[Lim Yan Peng Gary] Duke Increments

garylyp opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

garylyp committed:
Level 1. Greet, Echo, Exit

Changes made in this commit:
1) Initialize the basic I/O features of the programme.

garylyp committed:
Level 2. Add, List

Changes made in this commit:
1) Created a list to add and monitor input items.

garylyp committed:
Level 3. Mark as Done

Changes made in this commit:
1) Implemented feature to mark Task as done.

garylyp committed:
Level 4 ToDos, Events, Deadlines

Changes made in this commit:
1) Created the abstract Task class and various subclasses of Task.

Tasks are now distinct objects instead of just Strings. Each task
displays slightly different wordings on print.

garylyp committed:
Level 4. ToDos, Events, Deadlines (repost)

Changes made in this commit:
1) Introduced abstract Task class and implemented various subclasses.

Task are now distinct objects instead of just a String name as in the
third level.

garylyp committed:
Text UI Testing

Changes made in this commit:
1) Provided support to compare actual output against expected output

garylyp committed:
Level-5: Handle Errors

Changes made in this commit:
1) Included checks for various error inputs

Included a InvalidCommandException and EmptyDescriptionException to
check for faulty inputs.

garylyp committed:
Level 6: Delete

Changes made in this commit:
1) User can now delete a task from the existing list, provided the
   task index is valid in the first place.

garylyp committed:
Minor refactoring to task adding processes.

Changes made in this commit:
1) Separate the handling of actions and the adding of Tasks to
   taskList.

garylyp committed:
Level 7. Save

Changes made to this commit:
1) Allow taskList to be saved on disk while running program and
   loaded from disk on restart.

garylyp committed:
Level 8. Dates and Times

Changes made to this commit:
1) Restrict Deadline Task to take in only a specific date and time
   format as argument after the /by keyword.

garylyp committed:
Merge branch 'branch-Level-7'

garylyp committed:
Merge branch 'branch-Level-8' to branch 'master'

garylyp committed:
More OOP

Changes made in this commit:
1) Added TaskList, Storage, Parser and Ui classes to extract
   codes that are cluttered all over the main method in Duke.
2) Added Command and CommandCentre class to provide a more
   general way of adding new actions and their implementations
   to the chatbot.

garylyp committed:
Packages

Changes made in this commit:
1) Classified all classes into the relevant packages and updated
   access modifiers for methods affected.

garylyp committed:
JUnit Testing

Changes made in this commit:
1) Implemented Unit Testing with JUnit for specific cases in
   Parser and TaskList to handle different types of input.

Methods tested can be found in
src/test/java/package_name_of_method/NameOfClassBeingTestedTest.

The method names in the Test classes are in the following format
whatIsBeingTested_descriptionOfTestInputs_expectedOutcome().

If there are no test inputs. It is simply
whatIsBeingTested_expectedOutcome().

garylyp committed:
Merge branch 'gradle'

garylyp committed:
Package the App as a JAR file

Changes made to this commit:
1) Publish the app as jar file in build libraries.

garylyp committed:
Automate project builds using Gradle

Changes made in this commit:
1) Imported gradle project to the current project to enable
   build automation.

garylyp committed:
Minor bug fixes

Changes made to this commit:
1) Fix the problem of the program not exiting when bye command was
   entered.

garylyp committed:
Added JavaDoc

Changes made to this commit:
1) Added JavaDoc to the three main utils classes that lacked documentation.

garylyp committed:
Level 9. Find

Changes made to this commit:
1) Implemented "find" action and implementation to search for tasks
   corresponding to a particular keyword or phrase.

garylyp committed:
Merge branch 'branch-A-JavaDoc'

garylyp committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/utils/Ui.java

garylyp committed:
Added JavaDoc for methods added in branch-Level-9.

Changes made to this commit:
1) Added documentation to methods in branch-Level-9 after merging.

garylyp committed:
Tweak the code to comply with Google Java Style Guide.

Changes made in this commit:
1) Modified code style to abide by a common coding standard.

garylyp committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/utils/Ui.java

garylyp committed:
Updated JavaDocs

Changes made to this commit:
1) Added documentation to Ui class.

garylyp committed:
Level 10. GUI

Changes made in this commit:
1) Implemented GUI using Javafx library
2) Updated Parser such that it checks for valid date format for Events

fxml files and images were added to a newly created resources folder
to ensure the functioning of the javafx UI. As output must be constantly
managed as String in the dialog boxes, major refactoring has been made
to the Ui class such that every method that used to print a message to
the system now appends the message to a StringBuilder object before
being extracted as a String.

To be aligned with the Duke requirement, Event tasks can only accept
a specific Date format for the additional info like Deadline tasks.

Other notes:
The tick and cross symbols, along with the emoticon symbol found in
most warning messages, have been replaced with their unicode form as
their previous forms cannot be printed on the javafx ui.

garylyp committed:
Merge branch 'branch-A-JavaFX'

garylyp committed:
Resolve all CheckStyle Issues.

Changes made in this commit:
1) Resolve all checkstyle errors raised by the checker.

This includes the removal of indentation for 'case' for 'switch'
statements, the commenting of String objects that require unicode
characters (namely the tick, cross and emoticon) as well as clean-ups
to the documentation. No changes are made to the behaviour of the
program.

ShirleyWangxt commented on src/main/java/command/CommandCentre.java:
```diff
@@ -0,0 +1,53 @@
+     * A collection to store and map command names to their
+     * respective Command.
+     */
+    private HashMap<String, Command> commands;
```
I like how you use hashmap to represent commands which improves the efficiency of access.

ShirleyWangxt commented on src/main/java/main/Duke.java:
```diff
@@ -0,0 +1,183 @@
+
+public class Duke {
+
+    private static final String ROOT = "D:\\Gary\\Uni\\NUS\\1920SEM1\\CS2103T\\Practices\\duke";
```
Nice use of final constants.

ShirleyWangxt commented on src/main/java/exception/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,7 @@
+package exception;
+
+public class EmptyDescriptionException extends DukeException {
```
It is good that you use different classes to represent specific exceptions.

ShirleyWangxt commented on src/main/java/utils/Storage.java:
```diff
@@ -0,0 +1,139 @@
+            StringBuilder sb = new StringBuilder();
+            for (int i = 0; i < taskList.size(); i++) {
+                Task task = taskList.get(i);
+                if (task instanceof Todo) {
```
For each task instance, the steps to update data seems a bit long-winded and repetitive. Would it be better to construct the format in the respective subclasses instead of in the update method itself?

garylyp replied:
Good idea! I can create a method in Task for the formatting of the storage string so that the process is abstracted away from the Storage method. Thank you!

ShirleyWangxt commented on src/test/java/utils/ParserTest.java:
```diff
@@ -0,0 +1,110 @@
+package utils;
```
Your test classes seem pretty complete and specific. Nice job!

ShirleyWangxt commented on src/main/java/utils/Ui.java:
```diff
@@ -0,0 +1,162 @@
+    /**
+     * Prints a list of actions that can be used.
+     */
+    public void printHelpMessage() {
```
I like the help message function a lot! It serves a good purpose of introducing and assisting users of the functionality.

garylyp committed:
Rectify Junit test input for Parser method

With the recent updates in previous commits, Event detail can only
accept a specific date format as argument after the "/at" keyword.
The Junit test method input has has been corrected to attain a
successful result for a correct input format.

garylyp committed:
Use "assert" at certain key points in Ui, Storage and TaskList classes

garylyp committed:
Merge pull request #4 from garylyp/branch-A-Assertions

Use "assert" at certain key points in Ui, Storage and TaskList classes

garylyp committed:
Improve code quality throughout the project

Major code cleanups to all java files, especially in Parser and Ui
classes.

Parser used to be involved in I/O procedures but its job is
now strictly narrowed to validating user input and returning the data
to backend in the correct form. Parser is also now fully involved in
catching exceptions caused by invalid user input formats.

Ui no longer holds a singleton design as it is simply not necessary.

garylyp committed:
Merge pull request #5 from garylyp/branch-A-CodeQuality

Improve code quality throughout the project

ReignOfComputer commented:
Looks good to me.

garylyp committed:
Add "undo" commands to allow undo-ing of previous commands.

The "undo" command executes the opposite of the most recently
successfully executed command. It can be called repeatedly until
there is no more earlier commands executed in the current javafx
session.

garylyp committed:
Merge pull request #6 from garylyp/branch-C-Undo

Add "undo" commands to allow undo-ing of previous commands.

garylyp committed:
Add "sort" command to allow sorting of list

List can now be sorted by name, deadline, type of task or status, based
on user input. An additional argument "r" also renders sorting by
reversed order. For example, "sort deadline" sorts the task by deadline
in normal order while "sort name r" sorts the task by name in reversed
order

garylyp committed:
Merge pull request #7 from garylyp/branch-C-Sort

Add "sort" command to allow sorting of list

garylyp committed:
Add user guide to README.md

garylyp committed:
Set theme jekyll-theme-minimal

garylyp committed:
Generalize storage directory for release

Metadata is now stored in the root directory of the user. This allows
the jar file to restart from where it was left off after the previous
use.

garylyp committed:
Merge branch 'master' of https://github.com/garylyp/duke

garylyp committed:
Change storage directory to user.home

Improve help message for "help" command

garylyp committed:
Provide support for cross platform usage
